Apache Aries Activator needs updates to accommodate changes in OpenJ9 OpenJDK 17

I am testing with an early Semeru 17 development driver and I am getting the following error on server shutdown.
```
[9/4/25, 14:45:02:301 PDT] 00000046 com.ibm.ws.security.ready.internal.SecurityReadyServiceImpl  I CWWKS0009I: The security service has stopped.
[9/4/25, 14:45:03:600 PDT] 0000001b LogService-165-com.ibm.ws.org.apache.aries.jndi.core         E CWWKE0701E: FrameworkEvent ERROR org.osgi.framework.BundleException: Exception in org.apache.aries.jndi.startup.Activator.stop() of bundle com.ibm.ws.org.apache.aries.jndi.core.
	at org.eclipse.osgi.internal.framework.BundleContextImpl.stop(BundleContextImpl.java:916)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.stopWorker0(EquinoxBundle.java:1080)
	at org.eclipse.osgi.internal.framework.EquinoxBundle$EquinoxModule.stopWorker(EquinoxBundle.java:399)
	at org.eclipse.osgi.container.Module.doStop(Module.java:699)
	at org.eclipse.osgi.container.Module.stop(Module.java:557)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.decStartLevel(ModuleContainer.java:2153)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:2022)
	at org.eclipse.osgi.container.SystemModule.stopWorker(SystemModule.java:281)
	at org.eclipse.osgi.internal.framework.EquinoxBundle$SystemBundle$EquinoxSystemModule.stopWorker(EquinoxBundle.java:225)
	at org.eclipse.osgi.container.Module.doStop(Module.java:699)
	at org.eclipse.osgi.container.Module.stop(Module.java:557)
	at org.eclipse.osgi.container.SystemModule.stop(SystemModule.java:212)
	at org.eclipse.osgi.internal.framework.EquinoxBundle$SystemBundle$EquinoxSystemModule$1.run(EquinoxBundle.java:244)
	at java.base/java.lang.Thread.run(Thread.java:853)
Caused by: java.lang.IllegalStateException: Error setting field: no field found for type interface javax.naming.spi.ObjectFactoryBuilder
	at org.apache.aries.jndi.startup.Activator.swapStaticField(Activator.java:330)
	at org.apache.aries.jndi.startup.Activator.stop(Activator.java:269)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:896)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:1)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:757)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.stop(BundleContextImpl.java:888)
	... 13 more
 Event:org.osgi.framework.FrameworkEvent[source=com.ibm.ws.org.apache.aries.jndi.core_1.1.106.202509032037 [165]] 
```

Investigations lead to this [commit in openj9-openjdk-jdk17](https://github.com/ibmruntimes/openj9-openjdk-jdk17/commit/cc42041c2468b4ecb4bb24b342bdfe2d4dcc5839) which moved the `object_factory_builder` field of type`ObjectFactoryBuilder` from `NamingManager` to `NamingManagerHelper`, which is causing the above error to occur.

Judging by this piece of code in our [Activator overlay](https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.ws.org.apache.aries.jndi.core/src/org/apache/aries/jndi/startup/Activator.java#L92-L104), it seems like we've encountered this issue before with Java >= 20.

We'll need to update the overlay to accommodate this change in 17 as well. I don't see this error with the latest Semeru 17.0.16, but I think it will start occurring on the next release. I only see this now because I am working with a early development driver from Semeru team.

We may need to figure out a way to have the Aries JNDI factories register with the kernel set one so that it doesn't have to use these reflection tricks.

--- a/dev/com.ibm.ws.org.apache.aries.jndi.core/src/org/apache/aries/jndi/startup/Activator.java
+++ b/dev/com.ibm.ws.org.apache.aries.jndi.core/src/org/apache/aries/jndi/startup/Activator.java
@@ -19,8 +19,6 @@
 package org.apache.aries.jndi.startup;
 
 import java.lang.reflect.Field;
-import java.security.AccessController;
-import java.security.PrivilegedAction;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
@@ -28,7 +26,6 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
-import java.util.function.Supplier;
 
 import javax.naming.NamingException;
 import javax.naming.spi.InitialContextFactory;
@@ -88,34 +85,6 @@ public class Activator implements BundleActivator {
 
     private ObjectFactoryBuilder originalOFBuilder;
     private OSGiObjectFactoryBuilder ofBuilder;
-
-    //In Java20+ the ObjectFactoryBuilder static field is in the NamingManagerHelper class.
-    private final Class<?> objectFactoryBuilderHolder = new Supplier<Class<?>>() {
-        @Override
-        public Class<?> get() {
-            if (javaMajorVersion() >= 20) {
-                try {
-                    return Class.forName("com.sun.naming.internal.NamingManagerHelper");
-                } catch (ClassNotFoundException e) {
-                }
-            }
-            return NamingManager.class;
-        }
-    }.get();
-
-    private int javaMajorVersion() {
-        String version = AccessController.doPrivileged(new PrivilegedAction<String>() {
-            @Override
-            public String run() {
-                return System.getProperty("java.version");
-            }
-        });
-
-        String[] versionElements = version.split("\\D");
-
-        // Pre-JDK 9 java.version is 1.MAJOR.MINOR. Post-JDK 9 java.version is MAJOR.MINOR
-        return Integer.valueOf(versionElements[Integer.valueOf(versionElements[0]) == 1 ? 1 : 0]);
-    }
 
     public static Collection<ServiceReference<InitialContextFactoryBuilder>> getInitialContextFactoryBuilderServices() {
         return instance.icfBuilders.getReferences();
@@ -231,7 +200,7 @@ public class Activator implements BundleActivator {
                 } catch (IllegalStateException e) {
                     // use reflection to force the builder to be used
                     if (isPropertyEnabled(context, FORCE_BUILDER)) {
-                        originalOFBuilder = swapStaticField(objectFactoryBuilderHolder, ObjectFactoryBuilder.class, builder);
+                        swapObjectFactoryBuilderStaticField(builder);
                     }
                 }
                 ofBuilder = builder;
@@ -266,7 +235,7 @@ public class Activator implements BundleActivator {
             swapStaticField(NamingManager.class, InitialContextFactoryBuilder.class, originalICFBuilder);
         }
         if (ofBuilder != null) {
-            swapStaticField(objectFactoryBuilderHolder, ObjectFactoryBuilder.class, originalOFBuilder);
+            swapObjectFactoryBuilderStaticField(originalOFBuilder);
         }
 
         icfBuilders.close();
@@ -275,6 +244,23 @@ public class Activator implements BundleActivator {
         initialContextFactories.close();
 
         instance = null;
+    }
+
+    private void swapObjectFactoryBuilderStaticField(ObjectFactoryBuilder value) {
+        try {
+            // Try in NamingManager first.
+            swapStaticField(NamingManager.class, ObjectFactoryBuilder.class, value);
+        } catch (IllegalStateException ise) {
+            // The field may have been moved to NamingManagerHelper depending on the JDK version, so let's try there.
+            try {
+                Class<?> namingManagerHelperClass = Class.forName("com.sun.naming.internal.NamingManagerHelper");
+                swapStaticField(namingManagerHelperClass, ObjectFactoryBuilder.class, value);
+            } catch (ClassNotFoundException cnfe) {
+                // The NamingManagerHelper class does not exist, so the IllegalStateException may have been
+                // due to another reason, so let's throw the original exception.
+                throw ise;
+            }
+        }
     }
 
     private boolean isPropertyEnabled(BundleContext context, String key) {

